Implement combo colour serializing

Craftplacer opened the pull request:
Closes #9415

Craftplacer committed:
Initial commit

Craftplacer committed:
Add documentation for constructor

Craftplacer committed:
Handle additional null case

Craftplacer committed:
Temporary changes

Craftplacer committed:
Add required parameters and other various changes

Craftplacer committed:
Rename fields and make readonly

Craftplacer committed:
Merge remote-tracking branch 'upstream/master' into combo-colors

Craftplacer committed:
Change skin testing

Craftplacer committed:
Use existing field in Editor

Craftplacer committed:
Properly implement SkinConfiguration equality

Craftplacer committed:
Minor changes and improvements

Craftplacer committed:
Resolve review comments

Craftplacer committed:
Fix formatting

Craftplacer committed:
Remove Indent

smoogipoo commented on osu.Game/Beatmaps/BeatmapManager.cs:
```diff
@@ -194,14 +195,23 @@ protected override bool CheckLocalAvailability(BeatmapSetInfo model, IQueryable<
         /// <param name="beatmapContent">The <see cref="IBeatmap"/> content to write.</param>
-        public void Save(BeatmapInfo info, IBeatmap beatmapContent)
+        /// <param name="beatmapSkin">The beatmap <see cref="ISkin"/> content to write, or null if not to be changed.</param>
+        public void Save(BeatmapInfo info, IBeatmap beatmapContent, LegacyBeatmapSkin beatmapSkin = null)
```
Same here, this shouldn't take in a `LegacyBeatmapSkin`.

smoogipoo commented on osu.Game/Skinning/SkinConfiguration.cs:
```diff
@@ -45,8 +47,10 @@ public IReadOnlyList<Color4> ComboColours
+        public IDictionary<string, Color4> CustomColours { get; } = new SortedDictionary<string, Color4>();
 
-        public readonly Dictionary<string, string> ConfigDictionary = new Dictionary<string, string>();
+        public readonly SortedDictionary<string, string> ConfigDictionary = new SortedDictionary<string, string>();
```
Ignore. I think this is ok to stop unnecessary line changes.

frenzibyte commented on osu.Game/Screens/Edit/EditorChangeHandler.cs:
```diff
@@ -34,7 +36,8 @@ public class EditorChangeHandler : IEditorChangeHandler
         /// <param name="editorBeatmap">The <see cref="EditorBeatmap"/> to track the <see cref="HitObject"/>s of.</param>
-        public EditorChangeHandler(EditorBeatmap editorBeatmap)
+        /// <param name="beatmapSkin">The skin to track the inline skin configuration of.</param>
+        public EditorChangeHandler(EditorBeatmap editorBeatmap, LegacyBeatmapSkin beatmapSkin)
```
There won't be any issues, but by structure, `EditorBeatmap` is a wrapper for `IBeatmap` and is tracking any changes to the hit objects list, so for me that doesn't sound like the beatmap skin should be part of it, unless you plan for the `EditorBeatmap` to be more than just an `IBeatmap` wrapper.

frenzibyte replied:
Also for not taking a `LegacyBeatmapSkin`, initially it was taking an `ISkin` for beatmap skins, but I was against it for the fact that it could take literally any skin, not only beatmap skins. https://github.com/ppy/osu/pull/9957#discussion_r475192958

I've proposed having an `IBeatmapSkin` interface and using it in all the beatmap skin code paths.

bdach replied:
The legacy skin constraint should apply only to the legacy decoder. Again, I'm not sure what an interface would achieve here.

frenzibyte replied:
Yeah, I meant it to be a marker interface as a constraint to only skins created by `WorkingBeatmap`s, not any skin.

Craftplacer replied:
Any guidance on how to implement a skin property inside the EditorBeatmap? Otherwise, I'll try figuring this out by myself.

frenzibyte-alt replied:
Just add it as a normal property, what's the problem?

smoogipoo replied:
> Also for not taking a `LegacyBeatmapSkin`, initially it was taking an `ISkin` for beatmap skins, but I was against it for the fact that it could take literally any skin, not only beatmap skins. [#9957 (comment)](https://github.com/ppy/osu/pull/9957#discussion_r475192958)

@frenzibyte care to explain? Why should the beatmap _not_ be able to take in any skin?

frenzibyte-alt replied:
I've suggested it only as a constraint that `BeatmapManager.Save` for example only allows saving skins that come from `WorkingBeatmap`, but actually that sounds unnecessary... I take that back, and using `ISkin` in its place is correct, perhaps I rushed in thinking of suggesting it.

bdach commented:
@Craftplacer are you planning on continuing work on this?

Craftplacer commented:
> @Craftplacer are you planning on continuing work on this?

I did not forget about this, I'll get to resolving the comments this weekend.

Craftplacer committed:
Create a new BeatmapSetInfo when setting files

Craftplacer committed:
Merge remote-tracking branch 'upstream/master' into combo-colors

Craftplacer committed:
Use IBeatmapSkin

Craftplacer committed:
Safe checking against ComboColours instead of CustomColours

Craftplacer committed:
Remove if-cast

Craftplacer committed:
Scope down to IBeatmapSkin in EditorChangeHandler

Craftplacer committed:
Return true if both ComboColours are null

Craftplacer committed:
Fix xmldoc in LegacyBeatmapEncoder

Craftplacer committed:
Remove redundant call to TestResources

Craftplacer committed:
Make test load actual beatmap's skin configuration

Craftplacer committed:
Update osu.Game/Beatmaps/Formats/LegacyBeatmapEncoder.cs

Co-authored-by: Salman Ahmed <email@iisalman.me>

Craftplacer committed:
Remove empty argument list

Craftplacer committed:
Merge branch 'combo-colors' of https://github.com/Craftplacer/osu into combo-colors

Craftplacer committed:
Add BeatmapSkin to EditorBeatmap

Craftplacer committed:
Update LegacyBeatmapEncoderTest.cs

Craftplacer commented:
I've ran into an issue with the additional compare code inside `LegacyBeatmapEncoderTest`, currently we're passing the beatmap file to the `LegacySkin`/`LegacySkinDecoder` for handling. The issue being that the decoder accepts and collects any data thrown at it (e.g. beatmap data like hit objects) into its `ConfigDictionary`, making the tests fail.

I am currently stuck at this, asking myself on how to tackle this. I see commenting out problematic or unsupported data inside beatmap files not as a solution.

bdach commented on osu.Game/Beatmaps/BeatmapManager.cs:
```diff
@@ -194,14 +195,15 @@ protected override bool CheckLocalAvailability(BeatmapSetInfo model, IQueryable<
         /// <param name="info">The <see cref="BeatmapInfo"/> to save the content against. The file referenced by <see cref="BeatmapInfo.Path"/> will be replaced.</param>
         /// <param name="beatmapContent">The <see cref="IBeatmap"/> content to write.</param>
-        public void Save(BeatmapInfo info, IBeatmap beatmapContent)
+        /// <param name="beatmapSkin">The beatmap <see cref="ISkin"/> content to write, or null if not to be changed.</param>
```
I'm pretty sure the `or null if not to be changed` part is wrong.

bdach commented on osu.Game.Tests/Beatmaps/Formats/LegacyBeatmapEncoderTest.cs:
```diff
@@ -19,49 +22,115 @@
 
-            Assert.That(decodedAfterEncode.Serialize(), Is.EqualTo(decoded.Serialize()));
+            Assert.That(decodedAfterEncode.beatmap.Serialize(), Is.EqualTo(decoded.beatmap.Serialize()));
+            Assert.IsTrue(decodedAfterEncode.beatmapSkin.Configuration.Equals(decoded.beatmapSkin.Configuration));
```
a possible way out of this is to look one line above, and serialise the skins to JSON strings, after first ignoring the `ConfigDictionaries` that are being the issue here. that doesn't entirely resolve the test failures, though, as this PR does not handle `CustomColours`.

bdach commented on osu.Game.Tests/Beatmaps/Formats/LegacyBeatmapEncoderTest.cs:
```diff
@@ -19,49 +22,115 @@
+            public void Dispose() => beatmapData.Dispose();
+
+            public byte[] Get(string name)
+            {
+                if (name != beatmapName)
+                    return null;
+
+                byte[] buffer = new byte[beatmapData.Length];
+                beatmapData.Read(buffer, 0, buffer.Length);
+                return buffer;
+            }
+
+            public async Task<byte[]> GetAsync(string name)
+            {
+                if (name != beatmapName)
+                    return null;
+
+                byte[] buffer = new byte[beatmapData.Length];
+                await beatmapData.ReadAsync(buffer.AsMemory());
+                return buffer;
+            }
+
+            public Stream GetStream(string name)
+            {
+                if (name != beatmapName)
+                    return null;
+
+                beatmapData.Seek(0, SeekOrigin.Begin);
+                return beatmapData;
+            }
+
+            public IEnumerable<string> GetAvailableResources() => beatmapName.Yield();
+        }
+
+        private class TestLegacySkin : LegacySkin, IBeatmapSkin
+        {
+            public TestLegacySkin(Beatmap beatmap, IResourceStore<byte[]> storage, string fileName)
+                : base(new SkinInfo { Name = "Test Skin", Creator = "Craftplacer" }, storage, null, fileName)
+            {
+            }
```
is this all really necessary? can't you just `new LegacyBeatmapDecoder().Decode(stream)`?

bdach reviewed (dismissed):
aside from the entire `IBeatmapSkin` debacle (which I probably shouldn't have approved)

Craftplacer committed:
Revert changes done to SkinConfiguration and IHasCustomColours

Craftplacer committed:
Update xmldoc

Craftplacer committed:
Update LegacyBeatmapEncoder.cs

Craftplacer committed:
Cut down changes done to LegacyBeatmapEncoderTest

Craftplacer committed:
Revert left-over type change in SkinConfiguration

Craftplacer committed:
Remove set from properties in SkinConfiguration classes

I don't get why this wasn't resolved in the first place when this file was originally written.

*sigh*

Craftplacer committed:
Merge remote-tracking branch 'upstream/master' into combo-colors

Craftplacer committed:
Change types back

Craftplacer committed:
Update LegacyBeatmapEncoderTest.cs

bdach committed:
Revert unnecessary passing down of tuple in test

bdach commented:
made one cosmetic change here, I guess I would be ok with this being merged in this state myself

not entirely a fan of how `LegacyBeatmapDecoder` accepts an `ISkin` again, and the diff has another random change to `IHasComboColours` (setter removed - although I can kind of agree with that one), but frankly I don't really want to keep discussing trivialities on a branch like this which already has about as ten times as many commits as actually necessary

smoogipoo reviewed (approved):
Seems pretty fine to me!

smoogipoo committed:
Merge branch 'master' into combo-colors
